### PR TITLE
NeoClient refactor for data integrity and a more swifty approach

### PR DESCRIPTION
**Pull request by jaumevn**
Refoctors `NeoClient` implementing newest data validation approaches.

- This refactor includes new enums to handle server response; to ensure data validation and error propagation. (`NeoClientResult` and `NeoClientError`)

- Tests are updated to use latest `NeoClient` syntax for error handling and data validation.

### CODE REVIEW

**Comment by saltyskip** [comment]
Thank you for the PR @jaumevn, there are several things that I love in this PR. Guarding all forced unwraps is a huge win.

However I would like to further consider this style of error handling. I would consider it the Alamofire method of handling error propagation rather than the swifty way. To my knowledge there is no consensus in the Swift Community for passing errors up in an async call.

Admittedly the method that is currently being used in the repository is the "golang way." However you can still see the older methods that interop with objective C pass up separate data and error objects.

E.g
```
func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
```

Since there is no official Swift way, I hesitate to move away from the way apple does their asynchronous error handling, as this would be a big architectural change that would be the groundwork for all network calls as the project develops.

Any other swift devs want to chime in? @apisit perhaps

**Comment by jaumevn** [comment]
The use of enums for failure/success data handling is commonly used in most public SDKs in order to give better understanding of data/errors and also to provide an easy interface to work with.

Our main goal should be providing an easy-to-use SDK that:
- Handles all kind of errors/data.
- Handles all data integrity validation (that shouldn't be a developer task, that's the SDK task)
- Documents all SDK errors as well as success data using typed and validated data (that's why `NeoClientResult` enum is created)

Additionally, the use of generic enums make code more maintainable, easier to fix and update, and of course more testable. For instance, adding new methods or error to the SDK is easy. We will only need to add a new case in the `NeoClientError` enum and use the `NeoClientResult` with the new class/struct (or any kind of data) to be returned.

What do you think guys?

**Comment by saltyskip** [comment]
I can dig it

File: NeoSwift/Account.swift [hunk 1]
```diff
@@ -181,11 +187,24 @@ public class Account {
...
-                return
+        NeoClient.shared.getAssets(for: self.address, params: []) { result in
+            switch result {
+            case .failure(let error):
+                completion(nil, error)
+                break
+            case .success(let assets):
+                let payload = self.generateSendTransactionPayload(asset: asset, amount: amount, toAddress: toAddress, assets: assets)
+                NeoClient.shared.sendRawTransaction(with: payload, completion: { (result) in
+                    switch result {
```
Comment (saltyskip) [inline NeoSwift/Account.swift]: trailing closure syntax please

File: NeoSwift/Account.swift [hunk 2]
```diff
@@ -26,30 +26,36 @@ public class Account {
...
+        guard let wallet = GoNeowalletGenerateFromWIF(wif, &error) else { return nil }
         self.wif = wif
-        self.publicKey = (wallet?.publicKey())!
-        self.privateKey = (wallet?.privateKey())!
-        self.address = (wallet?.address())!
-        self.hashedSignature = (wallet?.hashedSignature())!
+        self.publicKey = wallet.publicKey()
+        self.privateKey = wallet.privateKey()
+        self.address = wallet.address()
+        self.hashedSignature = wallet.hashedSignature()
```
Comment (saltyskip) [inline NeoSwift/Account.swift]: Great! No crashes on bad key input :D

File: NeoSwift/NeoClient.swift [hunk 3]
```diff
@@ -37,11 +61,12 @@ public class NeoClient {
         self.seed = seed
     }
     
-    func sendRequest(_ method: RPCMethod, params: [Any]?, completion :@escaping ([String:AnyObject]?, Error?) -> Void) {
+    func sendRequest(_ method: RPCMethod, params: [Any]?, completion: @escaping (NeoClientResult<JSONDictionary>) -> ()) {
         guard let url = URL(string: seed) else {
-            completion(nil, "Invalid Seed" as? Error)
+            completion(.failure(NeoClientError.invalidSeed))
             return
```
Comment (saltyskip) [inline NeoSwift/NeoClient.swift]: can this be (.failure(.invalidSeed))

File: NeoSwift/NeoClient.swift [hunk 4]
```diff
@@ -52,181 +77,269 @@ public class NeoClient {
...
+            case .success(let response):
+                guard let txs = response["result"] as? [String] else {
+                    completion(.failure(NeoClientError.invalidData))
                     return
+                }
+                
+                let result = NeoClientResult.success(txs)
+                completion(result)
+                
+                break
```
Comment (saltyskip) [inline NeoSwift/NeoClient.swift]: why are we breaking here?

Comment (jaumevn) [reply NeoSwift/NeoClient.swift]: We can remove it only because it's the last case of the switch ;)

File: NeoSwift/NeoClient.swift (hunk 3 above)
Comment (jaumevn) [reply NeoSwift/NeoClient.swift]: Yep can be done, but then we will only return `NeoClientError` error type.

**Comment by apisit** [comment]
Thanks @jaumevn @saltyskip  this looks good!

**Comment by saltyskip** [comment]
Great!! One last change and we'll merge it in. We actually don't need breaks in any of our switch statements

No Implicit Fallthrough

In contrast with switch statements in C and Objective-C, switch statements in Swift do not fall through the bottom of each case and into the next one by default. Instead, the entire switch statement finishes its execution as soon as the first matching switch case is completed, without requiring an explicit break statement. This makes the switch statement safer and easier to use than the one in C and avoids executing more than one switch case by mistake.

Taken from https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/ControlFlow.html

**Comment by jaumevn** [comment]
Uoo! I didn't know that, we always learn something new ;)

**Pull request merged**